fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/DecorrelateValuesRule.java

```diff
@@ -0,0 +1,375 @@
+            // All the quantifiers are values boxes. We can't push them around as this would
+            // leave no children at all. So we exit here.
+            return;
+        }
+
+        //
+        // Create a translation map, and use it to translate the result value of this select as well as any predicates
+        //
+        final TranslationMap translationMap = createTranslationMapFromSelects(valuesById);
+        final Value newResultValue = selectExpression.getResultValue().translateCorrelations(translationMap, true);
+        final List<QueryPredicate> newPredicates = selectExpression.getPredicates().stream()
+                .map(predicate -> predicate.translateCorrelations(translationMap, true))
+                .collect(ImmutableList.toImmutableList());
+
+        //
+        // Push the values box into each child for which it is relevant.
+        //
+        final PushValuesIntoVisitor visitor = new PushValuesIntoVisitor(qunsByIdBuilder.build(), translationMap, call);
+        ImmutableList.Builder<Quantifier> newQuantifiersBuilder = ImmutableList.builderWithExpectedSize(selectExpression.getQuantifiers().size() - valuesById.size());
+        for (Quantifier qun : selectExpression.getQuantifiers()) {
+            if (valuesById.containsKey(qun.getAlias())) {
+                //
+                // This is one of the values boxes that we're pushing down. Omit this box entirely from the top-level select
+                //
+                continue;
+            }
+
+            boolean anyChanged = false;
+            ImmutableList.Builder<RelationalExpression> newExpressionsBuilder = ImmutableList.builderWithExpectedSize(qun.getRangesOver().getExploratoryExpressions().size());
+            for (RelationalExpression lowerExpression : qun.getRangesOver().getExploratoryExpressions()) {
+                if (correlatedToNone(lowerExpression, valuesById.keySet())) {
+                    newExpressionsBuilder.add(lowerExpression);
+                } else {
+                    anyChanged = true;
+                    newExpressionsBuilder.add(visitor.visit(lowerExpression));
+                }
+            }
+            if (anyChanged) {
+                // Some expression has changed. Create a new quantifier over the new expressions
+                Reference newRef = call.memoizeExploratoryExpressions(newExpressionsBuilder.build());
+                Quantifier newQun = qun.overNewReference(newRef);
+                newQuantifiersBuilder.add(newQun);
+            } else {
+                // No expression was correlated. Put in the original quantifier
+                newQuantifiersBuilder.add(qun);
+            }
+        }
+        call.yieldExploratoryExpression(new SelectExpression(newResultValue, newQuantifiersBuilder.build(), newPredicates));
+    }
+
+    @SuppressWarnings("PMD.AvoidBranchingStatementAsLastInLoop")
+    @Nullable
+    private SelectExpression findSelectForQuantifier(@Nonnull Quantifier.ForEach qun,
```

Comment:
We should try if possible to express this as a matcher. If we can't we should at least call this `findValuesForQuantifier(...)`